### PR TITLE
Retry Logic For AzureKeyVault based on status code

### PR DESCRIPTION
**Pull request by Deekshitha981**
**Task name**: <Name of changed or new pipeline task>

**Description**: <Describe your changes here>

**Documentation changes required:** (Y/N) <Please mark if documentation changes are required>

**Added unit tests:** (Y/N) <Please mark if unit tests were added or updated according changes>

**Attached related issue:** (Y/N) <Please add link to related issue here>

**Checklist**:
- [ ] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [ ] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts [hunk 1]
```diff
@@ -176,21 +179,41 @@ export class KeyVaultClient extends azureServiceClient.ServiceClient {
...
-            else if (response.statusCode == 400) {
-                return new azureServiceClientBase.ApiResult(tl.loc('GetSecretFailedBecauseOfInvalidCharacters', secretName));
-            }
-            else {
-                return new azureServiceClientBase.ApiResult(azureServiceClientBase.ToError(response));
-            }
-        }).then((apiResult: azureServiceClientBase.ApiResult) => callback(apiResult.error, apiResult.result),
-            (error) => callback(error));
+        this.invokeRequestWithRetry(httpRequest,secretName)
+    .then((apiResult: azureServiceClientBase.ApiResult) => callback(apiResult.error, apiResult.result))
```
Comment (starkmsu) [inline Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts:184]: Please fix indentation here and on the next ine

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts [hunk 2]
```diff
@@ -176,21 +179,41 @@ export class KeyVaultClient extends azureServiceClient.ServiceClient {
...
+    .catch(error => callback(error));
+
+    }
+
+    async invokeRequestWithRetry(httpRequest, secretName, retrysCount: number = 3, retryWait: number = 2000) {
+        return this.invokeRequest(httpRequest).then(async (response: webClient.WebResponse) => {
+             // Retry logic
+        let attempts = 0;
+                if (retrysCount > 0) {
```
Comment (starkmsu) [inline Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts:193]: Please fix indentation here and below

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts [hunk 3]
```diff
@@ -176,21 +179,41 @@ export class KeyVaultClient extends azureServiceClient.ServiceClient {
...
+        this.invokeRequestWithRetry(httpRequest,secretName)
+    .then((apiResult: azureServiceClientBase.ApiResult) => callback(apiResult.error, apiResult.result))
+    .catch(error => callback(error));
+
+    }
+
+    async invokeRequestWithRetry(httpRequest, secretName, retrysCount: number = 3, retryWait: number = 2000) {
+        return this.invokeRequest(httpRequest).then(async (response: webClient.WebResponse) => {
+             // Retry logic
+        let attempts = 0;
```
Comment (starkmsu) [inline Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts:192]: Please fix indentation here

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts [hunk 4]
```diff
@@ -176,21 +179,41 @@ export class KeyVaultClient extends azureServiceClient.ServiceClient {
...
+                    } catch (error) {
+                        console.error(tl.loc(`Attempt ${attempts + 1} failed with error: ${error}`));
+                        attempts++;
+                        console.log(tl.loc(`Retrying ...`));
+                            await new Promise(r => setTimeout(r, retryWait));
+                            return await this.invokeRequestWithRetry(httpRequest,secretName, retrysCount-1, retryWait)
+                        
+                    }
+                }
+                throw new Error(`Failed after ${attempts} attempts`);
```
Comment (starkmsu) [inline Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts:213]: attempts here will be always 1. Use retrysCount to track attempt number

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts [hunk 5]
```diff
@@ -176,21 +179,41 @@ export class KeyVaultClient extends azureServiceClient.ServiceClient {
...
-            }
-        }).then((apiResult: azureServiceClientBase.ApiResult) => callback(apiResult.error, apiResult.result),
-            (error) => callback(error));
+        this.invokeRequestWithRetry(httpRequest,secretName)
+    .then((apiResult: azureServiceClientBase.ApiResult) => callback(apiResult.error, apiResult.result))
+    .catch(error => callback(error));
+
+    }
+
+    async invokeRequestWithRetry(httpRequest, secretName, retrysCount: number = 3, retryWait: number = 2000) {
```
Comment (starkmsu) [inline Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts:189]: A type: retrysCount -> retriesCount

**Review by manolerazvan — APPROVED** [review]

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts [hunk 6]
```diff
@@ -5,6 +5,9 @@ import util = require("util");
 import tl = require('azure-pipelines-task-lib/task');
 import webClient = require("azure-pipelines-tasks-azure-arm-rest/webClient");
 
+let retriesCount: number;
```
Comment (starkmsu) [inline Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts:8]: Let's avoid declaring global variables - declare those as a part of method signature

File: _generated/AzureKeyVaultV1/operations/azure-arm-keyvault.ts [hunk 7]
```diff
@@ -5,6 +5,9 @@ import util = require("util");
 import tl = require('azure-pipelines-task-lib/task');
 import webClient = require("azure-pipelines-tasks-azure-arm-rest/webClient");
 
+let retriesCount: number;
```
Comment (starkmsu) [inline _generated/AzureKeyVaultV1/operations/azure-arm-keyvault.ts:8]: The same as above

**Review by starkmsu — APPROVED** [review]

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts (hunk 1 above)
Comment (Deekshitha981) [reply Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts]: Fixed the indentation

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts (hunk 3 above)
Comment (Deekshitha981) [reply Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts]: Fixed the indentation

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts (hunk 4 above)
Comment (Deekshitha981) [reply Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts]: modified the logic to 4-Retriescount

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts (hunk 5 above)
Comment (Deekshitha981) [reply Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts]: Changed to retriesCount

File: Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts (hunk 6 above)
Comment (Deekshitha981) [reply Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts]: modified the declaration as a part of method signature.

File: _generated/AzureKeyVaultV1/operations/azure-arm-keyvault.ts (hunk 7 above)
Comment (Deekshitha981) [reply _generated/AzureKeyVaultV1/operations/azure-arm-keyvault.ts]: modified the declaration as a part of method signature.

**Pull request merged**